### ISSUE TITLE
Wrong PBS flag Generation

### ISSUE THREAD

**Issue by roshant** [christopherwharrop/rocoto#47]
Hi,
I am using scheduler as "pbspro" in created xml script.
The scheduler on the cluster is PBSPRO.
Its not allowing "-l select=<nchunks>:ncpus=<no. of cores>" flag in <native> tag. Its always wants nodes/cores tag to be there.
When I am using:
<nodes>2:ppn=36</nodes>
tag in xml script its evaluating to below PBS flag:
-l select=2:mpiprocs=36:ncpus=16.
Which is resulting in 16 cores per node though on the cluster there are 36 cores per node.
Please help in sorting out this error. I want to use PBS pro with following flag:
-l select=2:ncpus=36:vntype=cray_compute -l place=scatter

Regards,
Roshan

**Comment by christopherwharrop** [comment]
Sorry for the delay.  Can you please provide me the exact XML you are using (either the `<cores>` or `<nodes>` tag)?  I want to be sure that I'm looking at the same thing you are.  PBSPro, as far as I know, forces you to request specific number of nodes.  This makes it hard for Rocoto to translate the  `<cores>` tag into the proper request.  A `<nodesize>` tag was added to assist with that.  If you specify `<cores>`, you also should specify `<nodesize>` as well.  If you don't, Rocoto will run a pbs command to try to guess the node size (and hyperthreads can cause this guess to be wrong).  It should print a warning when it does this.  The combination of `<cores>` and `<nodesize>` allows Rocoto to compute the correct distribution of nodes/cores in the -l select request.

If it is still doing the wrong thing, even with that, then there is probably a bug that needs fixing.

Unfortunately, though, the `vntype=cray_compute` specification isn't possible with Rocoto.  And, unfortunately, the `-l place=scatter` isn't possible either, as there is no generic tag that translates to that.  For both of these, you'll have to use the `<custom>` tag to hard code your precise resource request directly.  But, this probably won't work right now because I believe Rocoto currently insists that either `<cores>` or `<nodes>` is present, which you can't use.  In light of the complexity that node properties introduces into the syntax of resource requests, that requirement should probably be dropped, and that is easy to do.  I am happy to make that modification.

**Comment by roshant** [comment]
Hello Christopher,

Thank you for the reply for the posted issue.
Sorry for the late reply.
The XML script I am using is attached to this email. The PBS script I want
to execute is also attached along with this email. I will suggest, there
should be a tag ( resource request tag) like <cores> and <nodes> which is
having higher precedence if its there in the XML file. Or there should be
some mechanism which will translate User job script when provided
automatically by ROCOTO.
The "select" flag is computing properly now, when I am using <cores> and
<nodesize> flag .
The issue was coming when I was using <nodes>2:ppn=36</nodes> tag , which
was resulting into :
-l select=2:mpiprocs=36:ncpus=16

Please notify, If ROCOTO able to parse :

-l select=2:ncpus=36:vntype=cray_compute -l place=scatter

flag in XML script.

Thanks and Best Regards,
Roshan

**Comment by christopherwharrop** [comment]
I have not found a great way to solve this with the existing <nodes> tag support.  So, I have added the capability to use <native> as an alternative to <nodes> such that the exact option required can be specified.  It will result in the workflow being non-portable across batch systems, but it should allow any option to be specified as necessary.  This was added in PR #61

**Comment by christopherwharrop-noaa** [comment]
I am reopening this issue.  My previous thinking on this was wrong.  I think I have a much better idea how to do this.

**Comment by christopherwharrop-noaa** [comment]
@roshant , I have made some additional fixes. If you are still interested in this, can you please check out the bugfix/nodes-consumables branch and see if it gives you what you want?  You should be able to do:

```
<nodes>2:ppn=36:vntype=cray_compute </nodes>
<native>-l place=scatter</native>
```

and it should produce:

```
#PBS -l select=2:mpiprocs=36:vntype=cray_compute:ncpus=36
#PBS -l place=scatter
```

**Comment by roshant** [comment]
Hi Christopherwharrop/Rocoto,

Thank you for the reply and bug fixes and extremely sorry for the delayed
reply. I am currently out of station and will start working on it  from
July 1, 2019. Will let you know the made bug fixes are working or not. Will
contact you in case of any difficulties.

Regards,
Roshan

**Comment by christopherwharrop-noaa** [comment]
@roshant , Please let me know when you have had a chance to try the `bugfix/nodes-consumables` branch.  If you are successful, I will merge and release it.

**Comment by roshant** [comment]
Hello Christopherwharrop/Rocoto,
The made bug fixes are successful. Its working fine now. I have made the
following changes in the xml file.
Initially it was failing as the <nodesize> tag was missing and was taking
16 cores as default. So I added it and it started working fine.

<nodes>2:ppn=36:vntype=cray_compute</nodes>
<nodesize> 36 </nodesize>
<native> -l place=scatter </native>

Thank you for the bug fixes.

Regards,
Roshan

**Comment by christopherwharrop-noaa** [comment]
Thank you for your feedback.  I have merged this and released a bugfix version 1.3.1 that includes this fix.  You can download and use 1.3.1 as needed.
